### PR TITLE
Jupyter UI that manages Notebook CRs

### PR DESCRIPTION
**Pull request by kimwnasptd**
### Description
This PR provides the initial version of the new Web UI to natively spawn Jupyter Notebooks in a Kubeflow environment. The new Jupyter UI aims to replace the existing `jupyter` component and disengage Kubeflow from the JupyterHub API.

Please note that the new Jupyter UI is still under development, so there will be various fixes/improvements to its appearance and functionality until it’s ready to be merged. This WIP PR will also be updated soon to support the existing 3rd party alternative Jupyter UIs and not just the default one.

To quickly test the new Jupyter UI in your K8s cluster we provide a series of deployment steps. We have already tested those on Minikube, bare metal K8s and GKE. Of course, we welcome feedback and encourage discussion on issues that might arise.

### Closes
https://github.com/kubeflow/kubeflow/issues/1995
https://github.com/kubeflow/kubeflow/issues/2156
https://github.com/kubeflow/kubeflow/issues/2232

### Depends
https://github.com/kubeflow/kubeflow/pull/2349
Τhis PR has been updated and now depends on #2456 

### Main features of the new Jupyter UI

- Designed as a native K8s client for managing custom Notebook resources - no intermediate APIs used
- Leverages @kam’s Notebook CRD to spawn Notebook Pods via the respective Controller, instead of depending on JH KubeSpawner
- Written in Python Flask + plain HTML/jQuery/Material Design Lite. This stack allows reusability of already implemented components and delivery of a frontend that shares similar aesthetic with the central dashboard of Kubeflow.
- Currently uses a privileged ServiceAcccount that is allowed to manage Notebook CRs along with PVCs in all namespaces

### Future/Related Work
Integration with the new login UI component, as well as @kam’s Profiles when multiple (read/write) ServiceAccounts are supported. The new Jupyter web application will use its own privileged ServiceAccount to dispatch all API requests. In a later iteration it will be using [Impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation).

- @jlewi We need to investigate which is the best way to e2e test the new Jupyter component, meaning both user interface and backend. Related to:
https://github.com/kubeflow/kubeflow/issues/2174
https://github.com/kubeflow/kubeflow/pull/2333
https://github.com/kubeflow/kubeflow/pull/2323
- @kam @jlewi We need to implement a watcher on K8s Notebooks so that we retrieve reliable, real-time on the status of each created CR. A good starting point could be to map the status of the Jupyter Notebook to Pod lifecycle states. For this, we might also need to extend the Notebook CRD. This will definitely improve UX and increase visibility on deployed Notebooks.
- @jlewi Update the `centraldashboard` component of Kubeflow to replace “JupyterHub” link with “Notebooks” and route to the new Jupyter UI. As a side change, we would also need to update the corresponding public Docker image on gcr.io for 0.5.

### Deployment Steps
(Minor tweaks might be needed depending on your development environment)

1. Clone the Arrikto Kubeflow repo and checkout the `feature-jupyter-web-app` branch
`git clone -b feature-jupyter-web-app https://www.github.com/arrikto/kubeflow.git`
2. Build the new Jupyter UI Docker image locally on the host running K8s
`docker build -t jupyter-web-app:v0.1 kubeflow/components/jupyter-web-app`
3. Initialize the ksonnet application
`kubeflow/scripts/kfctl.sh init kf`
4. Generate and apply existing KF components via ksonnet
`cd kf && ../kubeflow/scripts/kfctl.sh generate all && ../kubeflow/scripts/kfctl.sh apply all`
5. Instruct the `jupyter-web-app` KF component to use the image you previously built
`cd ks_app && ks param set jupyter-web-app image jupyter-web-app:v0.1`
6. Apply the `jupyter-web-app` KF component
`ks apply default -c jupyter-web-app`
7. Configure your K8s cluster networking and make sure that the Kubeflow central Dashboard is reachable from outside the cluster
8. Point your browser to `http://<KUBEFLOW_IP>/jupyter/` (trailing slash). You should be able to access the dashboard of the new Jupyter UI.
9. Follow the instructions located at `kubeflow/components/jupyter-web-app/README.md` to create, connect to and delete Jupyter Notebooks.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2357)

### CODE REVIEW

**Comment by pdmack** [comment]
Traveling in Europe but might be able to look tomorrow

**Comment by jlewi** [comment]
Related to testing
kubeflow/testing#288 about getting selenium working to write UI based tests.

**Comment by jlewi** [comment]
@kimwnasptd This looks great to me. I mostly looked at the python code.

Hopefully someone with more frontend experience like @avdaredevil can take a look at the front end stuff.

Is there specific issues or questions that need to be resolved? In other words what's still in progress? If nothing and its ready to review please go ahead and remove WIP from the title to signal to reviewers that its read to be reviewed.

If you want to block it from being merged you can put a hold on it.

**Comment by jlewi** [comment]
> Yes. currently we've added extra privileges to the webapp's service account to list/edit/get/delete Notebooks, PVCs and Namespces. Then the webapp uses this sa's token for any call to the api server. Is this enough for now or should it use impersonation as described in the design doc?

Yes this is fine.

> Also there are some last bits we need to work on the UI to have the same functionality with the current implementation. We'll be ready to remove the WIP label within the week.

What is this functionality? Can we do it in a follow on PR and get the initial version merged?

**Comment by jlewi** [comment]
@kimwnasptd How's it going? Any idea on when you'll have the next iteration of this PR?

File: components/jupyter-web-app/jupyter/static/js/notebooks.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  });
+}
+
+// Search Bar
+$('#search-nb').bind("enterKey",function(e){
+  var ns = this.value;
+  updateNotebooksInNamespace(ns);
+
+  // Change the function for the CREATE NOTEBOOK button
+  $("#create-nb-btn").attr("onclick", "createNotebook('" + ns + "')")
```
Comment (avdaredevil) [inline components/jupyter-web-app/jupyter/static/js/notebooks.js]: Perhaps go with:

```js
$("#create-nb-btn").click(_ => createNotebook(ns))
```
OR
```js
$("#create-nb-btn").click(createNotebook.bind(0, ns))
```

File: components/jupyter-web-app/jupyter/static/js/notebooks.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  $("#ns-select").text("")
+}
+
+// Update the Notebooks when a Namespace is selected
+$("#ns-select").on("change", function() {
+  var ns = this.value;
+  updateNotebooksInNamespace(ns);
+
+  // Change the function for the CREATE NOTEBOOK button
+  $("#create-nb-btn").attr("onclick", "createNotebook('" + ns + "')")
```
Comment (avdaredevil) [inline components/jupyter-web-app/jupyter/static/js/notebooks.js]: Same as review above

**Pull request merged**